components/resource_control/src/resource_group.rs

```diff
@@ -0,0 +1,400 @@
+    use yatp::queue::Extras;
+
+    use super::*;
+
+    fn new_resource_group(
+        name: String,
+        is_ru_mode: bool,
+        read_tokens: f64,
+        write_tokens: f64,
+    ) -> ResourceGroup {
+        let mut group = ResourceGroup::new();
+        group.set_name(name);
+        let mode = if is_ru_mode {
+            GroupMode::RuMode
+        } else {
+            GroupMode::NativeMode
+        };
+        group.set_mode(mode);
+        if is_ru_mode {
+            let mut ru_setting = GroupRequestUnitSettings::new();
+            ru_setting.mut_r_r_u().set_tokens(read_tokens);
+            ru_setting.mut_w_r_u().set_tokens(write_tokens);
+            group.set_r_u_settings(ru_setting);
+        } else {
+            let mut resource_setting = GroupResourceSettings::new();
+            resource_setting.mut_cpu().set_tokens(read_tokens);
+            resource_setting.mut_io_write().set_tokens(write_tokens);
+            group.set_resource_settings(resource_setting);
+        }
+        group
+    }
+
+    #[test]
+    fn test_resource_group() {
+        let mut resource_manager = ResourceGroupManager::default();
+        resource_manager.total_ru_quota = 10000.0;
+
+        let group1 = new_resource_group("TEST".into(), true, 100.0, 100.0);
+        resource_manager.add_resource_group(group1);
+
+        assert!(resource_manager.get_resource_group("test1").is_none());
+
+        let group = resource_manager.get_resource_group("test").unwrap();
+        assert_eq!(
+            group.value().get_r_u_settings().get_r_r_u().get_tokens(),
+            100.0
+        );
+        drop(group);
+        assert_eq!(resource_manager.resource_groups.len(), 1);
+
+        let group1 = new_resource_group("Test".into(), true, 200.0, 100.0);
+        resource_manager.add_resource_group(group1);
+        let group = resource_manager.get_resource_group("test").unwrap();
+        assert_eq!(
+            group.value().get_r_u_settings().get_r_r_u().get_tokens(),
+            200.0
+        );
+        drop(group);
+        assert_eq!(resource_manager.resource_groups.len(), 1);
+
+        let group2 = new_resource_group("test2".into(), true, 400.0, 200.0);
+        resource_manager.add_resource_group(group2);
+        assert_eq!(resource_manager.resource_groups.len(), 2);
+
+        let resouce_ctl = resource_manager.derive_controller("test_read".into(), true);
+        assert_eq!(resouce_ctl.resource_consumptions.len(), 3);
+
+        let group1 = resouce_ctl.resource_group("test".as_bytes());
+        assert_eq!(group1.weight, 500);
+        let group2 = resouce_ctl.resource_group("test2".as_bytes());
+        assert_eq!(group2.weight, 250);
+        assert_eq!(group1.current_vt(), 0);
+
+        let mut extras1 = Extras::single_level();
+        extras1.set_metadata("test".as_bytes().to_owned());
+        assert_eq!(resouce_ctl.priority_of(&extras1), 25_000);
+        assert_eq!(group1.current_vt(), 25_000);
```

Comment:
Does that mean `vt` increased 25ms? will it weight too large?

Reply:
We don't care much about the vt's factor, we just care about the proportion between different groups.